Feat/#7 클린 아키텍쳐 적용

heyoni opened the pull request:
## 클린아키텍쳐 적용
- #8
- #7
- #6

## TODO
- Git Action 테스트 자동확
- docker

heyoni committed:
feat: 코드리뷰 반영

- password_check는 FE에서 체크하도록 변경
- password_check 삭제 후 test 코드 수정

heyoni committed:
feat: 클린 아키텍쳐 적용

- 클린 아키텍쳐 의존성 역전(DIP) 적용
  - 기존에는 endpoint에서 DB에 바로 접근하였는데, Repository를 이용하여 의존성 변경
  - Repository는 Abstract를 이용하여 DIP 패턴 적용
  - Service 패턴을 이용하여 endpoint와 Repository를 분리

heyoni committed:
feat: #6 mock을 이용한 테스트 코드 작성

#6

heyoni committed:
feat: test 코드 추가

heyoni committed:
feat: log 추가

heyoni committed:
feat: github action 추가

- 테스트코드 자동 실행 추가(github action)

f-lab-zion commented on .github/workflow/test_before_commit.yml:
```diff
@@ -0,0 +1,21 @@
+      - uses: actions/setup-python@v2
+        with:
+          python-version: "3.10"
+      - run: |
+          touch .env
+          echo SQLALCHEMY_DATABASE_URL=${{ secrets.SQLALCHEMY_DATABASE_URL }} >> .env
+          cat .env
+      - run: python -m venv env
+      - run: source env/bin/activate
+      - run: pip install -r requirements.txt
+      - run: python -m pytest
```
[actions:python-app](https://github.com/f-lab-edu/What-cat-eat/new/develop?filename=.github%2Fworkflows%2Fpython-app.yml&workflow_template=ci%2Fpython-app
)
위 링크를 참고하셔서 pip upgrade, pip install requirements, flake8, black, pytest를 순차적으로 테스트하도록 수정해보세요

heyoni replied:
넵 :) env 부분은 docker 설정하면서 변경하겠습니다!(다음 PR 예정입니다.)

f-lab-zion commented on repositories/UserRepository.py:
```diff
@@ -0,0 +1,83 @@
+
+class AbstractRepository(abc.ABC):
+    @abc.abstractmethod
+    def create(self, user: User):
```
범용적인 abstract method에 User라는 클래스는 어울리지 않습니다

f-lab-zion commented on repositories/UserRepository.py:
```diff
@@ -0,0 +1,83 @@
+            return None
+        return user
+
+    def get_value(self, user: User, nickname: str = None, user_id: str = None) -> User:
```
nickname 검색과 user_id 검색을 하나로 할 필요가 있을까요? 두개를 or 조건을 걸면 어떤게 먼저 나올지 예측이 될까요?

heyoni replied:
이 함수는 nickname, user_id 중복체크를 위해 작성했습니다. 함수명을 바꿀까요? 아니면 nickname혹은 user_id를 받게 되면 user(혹은 False)를 반환하는 함수로 변환할까요?

mihosang replied:
a라는 유저의 user id와 b라는 유저의 nickname 이 이 함수의 파라미터로 온다면 어떤 동작이 발생할지 예측이 어렵게 되겠죠
이런 경우라면 user id를 받는 함수와 nickname을 받는 함수로 나누는게 의미가 명확하고 side effect 여지를 줄여줍니다

f-lab-zion replied:
a라는 유저의 user id와 b라는 유저의 nickname 이 이 함수의 파라미터로 동시에 입력되면 어떤 동작이 발생할지 예측이 어렵게 되겠죠
이런 경우라면 user id를 받는 함수와 nickname을 받는 함수로 나누는게 의미가 명확하고 side effect 여지를 줄여줍니다

heyoni replied:
잘 이해가 가지 않습니다..ㅜㅜ a유저의 user_id랑 b라는 nickname은 결국 get_value를 호출하는 쪽에서 잘못 입력한 것이라 호출한 쪽의 잘못이라 생각되는데요.
만약 나눈다고 해도 똑같이 a유저의 user_id, b유저의 nickname을 입력하게 된다면 똑같은 오류가 나는게 아닌가요?!

heyoni replied:
이 코드리뷰는 **db를 조회하거나 함수의 역할을 넣어줄 때 두루뭉술하게 하는 것보다는 딱 집어서 해주는게 좋고, 여러가지 파라미터를 넣는 것보다는 명확하게 하나만 하는 게 side effect를 줄일 수 있다.**
라고 이해했고 코드도 알맞게 변경했습니다.

f-lab-zion commented on services/UserServices.py:
```diff
@@ -0,0 +1,63 @@
+        return user
+
+    def create(self, user_body: UserCreate) -> User:
+        validate_password(user_body)
```
validate_password 의 parameter로 user_body를 넘길게 아니라 user_body.password를 넘기는게 역할이 더 명확해보입니다
validate_password 함수에서도 Exception 을 raise 하기 보다는 True/False의 Boolean return하여 create, update 함수에서 return value에 따라 분기하는게 좋아보입니다

heyoni replied:
넵 확인했습니다!

f-lab-zion commented on services/UserServices.py:
```diff
@@ -0,0 +1,63 @@
+        if not user_exist:
+            raise HTTPException(status_code=404, detail="사용자를 찾을 수 없습니다.")
+
+        validate_password(user_body)
```
위와 동일

f-lab-zion commented on api/v1/endpoints/user.py:
```diff
@@ -17,41 +8,22 @@
-        )
-
-    return create_user(user_create=_user_create, db=db)
+def create_user(user: UserCreate, userService: UserService = Depends()):
```
UserCreate 에는 user_id만 있는데 password, nickname은 어디서 받나요?

heyoni replied:
UserBase를 상속받고 있습니다. 변경해야 할까요?
```python
class UserBase(BaseModel):
    id: Optional[int] = Field(primary_key=True)
    nickname: str = Field(min_length=1, max_length=15)
    password: str


class UserCreate(UserBase):
    user_id: str = Field(min_length=1, max_length=15)
```

f-lab-zion replied:
아 이건 제가 놓쳤네요 확인했습니다 변경하실 필요 없습니다

f-lab-zion commented on api/v1/endpoints/user.py:
```diff
@@ -17,41 +8,22 @@
-        raise HTTPException(status_code=404, detail="사용자를 찾을 수 없습니다.")
-    update_user(user_update=_user_update, id=id, db=db)
-    return user
+def update_user(id: int, user_body: UserUpdate, userService: UserService = Depends()):
```
UserUpdate 도 비어있는거 같습니다. model 쪽 확인 부탁드려요

heyoni replied:
여기도 UserBase를 상속받고 있습니다.
```python
class UserUpdate(UserBase):
    pass
```

f-lab-zion reviewed (changes requested):
클린 아키텍처 적용이 좋아보입니다. 리뷰하기도 상당히 편해졌네요 
디테일한 부분 comment 드렸으니 살펴보시기 바랍니다.

heyoni committed:
feat: 코드리뷰 반영

- 추상 메서드에서 파라미터로 받아오는 구체 클래스명을 추상화
- 가입시 중복 체크용 get_value 함수를 하나의 컬럼만 받아오는 함수로 변경
- validate_password 파라미터 변경, 내부 exception 을 service로 이동
- flake8, black 적용

heyoni committed:
잘못 올라간 파일 삭제

heyoni committed:
사용하지 않는 crud 기능 삭제

heyoni committed:
삭제된 파일 재업로드

f-lab-zion commented on repositories/UserRepository.py:
```diff
@@ -0,0 +1,102 @@
+    def get_user_by_user_id(self, user: User, user_id: str = None) -> User:
+        user = self.db.query(User).filter(User.user_id == user_id).first()
+        if not user:
+            return False
```
return type을 User class로 했으면 여기서도 `return None`이 더 적절할거 같습니다

f-lab-zion commented on repositories/UserRepository.py:
```diff
@@ -0,0 +1,102 @@
+    def get_user_by_nickname(self, user: User, nickname: str = None) -> User:
+        user = self.db.query(User).filter(User.nickname == nickname).first()
+        if not user:
+            return False
```
위와 동일

f-lab-zion commented on schema/user.py:
```diff
@@ -25,20 +25,22 @@ class Config:
-        raise HTTPException(status_code=400, detail="비밀번호가 일치하지 않습니다.")
+def validate_password(password) -> dict:
+    if len(password) < 8:
+        return {"success": False, "status_code": 400, "error": "비밀번호 길이가 너무 짧습니다."}
```
dict로 리턴하기 보다는 pydantic model을 쓰면 key를 누락하는 등 실수를 줄일 수 있습니다

f-lab-zion commented on repositories/UserRepository.py:
```diff
@@ -0,0 +1,102 @@
+            return None
+        return user
+
+    def get_user_by_user_id(self, user: User, user_id: str = None) -> User:
```
파라미터 중 user: User는 불필요하지 않나요? 어차피 query() 에 들어가는 User class는 상황에 따라 바뀌는 변수가 아닌거 같습니다만

f-lab-zion commented on repositories/UserRepository.py:
```diff
@@ -0,0 +1,102 @@
+            return False
+        return user
+
+    def get_user_by_nickname(self, user: User, nickname: str = None) -> User:
```
위와 동일

f-lab-zion reviewed (commented):
👍 몇가지 코멘트 남깁니다

heyoni committed:
feat: 코드리뷰 반영

- User타입을 가져오는 경우 return False 에서 None으로 변경
- dict에서 pydantic 모델로 변경
- 불필요한 파라미터 삭제
